pkg/controller/kubernetes/containerregistry.go

```diff
@@ -0,0 +1,203 @@
+	errGetContianerRegistirySubscription = "cannot get DOContainerRegistery subscrioption"
+
+	errContianerRegistiryCreateFailed = "creation of DOContainerRegistery resource has failed"
+	errContianerRegistiryDeleteFailed = "deletion of DOContainerRegistery resource has failed"
+	errContianerRegistiryUpdate       = "cannot update managed DOContainerRegistery resource"
+
+	subscriptionOutDated = "subscription is not up to date"
+)
+
+// SetupDOContainerRegistry adds a controller that reconciles DOContainerRegistry managed
+// resources.
+func SetupDOContainerRegistry(mgr ctrl.Manager, l logging.Logger) error {
+	name := managed.ControllerName(v1alpha1.DOContainerRegistryGroupKind)
+
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		For(&v1alpha1.DOContainerRegistry{}).
+		Complete(managed.NewReconciler(mgr,
+			resource.ManagedKind(v1alpha1.DOContainerRegistryGroupVersionKind),
+			managed.WithExternalConnecter(&containerRegistryConnector{kube: mgr.GetClient()}),
+			managed.WithReferenceResolver(managed.NewAPISimpleReferenceResolver(mgr.GetClient())),
+			managed.WithConnectionPublishers(),
+			managed.WithInitializers(managed.NewDefaultProviderConfig(mgr.GetClient())),
+			managed.WithLogger(l.WithValues("controller", name)),
+			managed.WithRecorder(event.NewAPIRecorder(mgr.GetEventRecorderFor(name)))))
+}
+
+type containerRegistryConnector struct {
+	kube client.Client
+}
+
+func (c *containerRegistryConnector) Connect(ctx context.Context, mg resource.Managed) (managed.ExternalClient, error) {
+	token, err := do.GetAuthInfo(ctx, c.kube, mg)
+	if err != nil {
+		return nil, err
+	}
+	client := godo.NewFromToken(token)
+	return &containerRegistryExternal{Client: client, kube: c.kube}, nil
+}
+
+type containerRegistryExternal struct {
+	kube client.Client
+	*godo.Client
+}
+
+func (c *containerRegistryExternal) Observe(ctx context.Context, mg resource.Managed) (managed.ExternalObservation, error) {
+	cr, ok := mg.(*v1alpha1.DOContainerRegistry)
+	if !ok {
+		return managed.ExternalObservation{}, errors.New(errNotContianerRegistiry)
+	}
+
+	if meta.GetExternalName(cr) == "" {
+		return managed.ExternalObservation{
+			ResourceExists: false,
+		}, nil
+	}
+
+	observed, response, err := c.Registry.Get(ctx)
+	if err != nil {
+		return managed.ExternalObservation{}, errors.Wrap(do.IgnoreNotFound(err, response), errGetContianerRegistiry)
+	}
+
+	subscription, response, err := c.Registry.GetSubscription(ctx)
+	if err != nil {
+		return managed.ExternalObservation{}, errors.Wrap(do.IgnoreNotFound(err, response), errGetContianerRegistirySubscription)
+	}
+
+	currentSpec := cr.Spec.ForProvider.DeepCopy()
+	dok8s.RegistryLateInitializeSpec(&cr.Spec.ForProvider, *observed)
+	if !cmp.Equal(currentSpec, &cr.Spec.ForProvider) {
+		if err := c.kube.Update(ctx, cr); err != nil {
+			return managed.ExternalObservation{}, errors.Wrap(err, errContianerRegistiryUpdate)
+		}
+	}
+
+	cr.Status.AtProvider = v1alpha1.DOContainerRegistryObservation{
```

Comment:
This is my fault since it's in the other controllers as well, but can we move this into `pkg/clients/kubernetes/containerregistry.go` as something like `GenerateContainerRegistryObservation(observed, subscription)` and have it return the `DOContainerRegistryObservation`?

I'm going to be refactoring a lot of this code and that's one of the first things I'm going to do, so if not that's fine.